Split Object.defineProperty() example

Josh-Cena opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

Fix #23172. Also ran Prettier in https://github.com/mdn/content/pull/23266/commits/b33e05392aecd6804c4f7b4d26f51d85053cc06a. Check https://github.com/mdn/content/pull/23266/commits/a846182e4497cb77860dc5c2c743f2c12df6942c for content changes.

### Motivation

<!-- ❓ Why are you making these changes and how do they help readers? -->

### Additional details

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

Josh-Cena committed:
Format

Josh-Cena committed:
Split Object.defineProperty() example

estelle commented on files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:
```diff
@@ -30,33 +28,21 @@ Object.defineProperty(obj, prop, descriptor)
-{{jsxref("Operators/delete", "deleted", "", 1)}}. This method allows these extra details
-to be changed from their defaults. By default, properties added using
-`Object.defineProperty()` are not writable, not enumerable, and not configurable.
+`Object.defineProperty()` allows a precise addition to or modification of a property on an object. Normal property addition through [assignment](/en-US/docs/Web/JavaScript/Reference/Operators/Assignment) creates properties which show up during property enumeration ({{jsxref("Statements/for...in", "for...in")}}, {{jsxref("Object.keys()")}}, etc.), whose values may be changed, and which may be {{js ...
```
```suggestion
`Object.defineProperty()` allows a precise addition to or modification of a property on an object. Normal property addition through [assignment](/en-US/docs/Web/JavaScript/Reference/Operators/Assignment) creates properties which show up during property enumeration ({{jsxref("Statements/for...in", "for...in")}}, {{jsxref("Object.keys()")}}, etc.), whose values may be changed and which may be {{jsxref("Operators/delete", "deleted", "", 1)}}. This method allows these extra details to be changed from their defaults. By default, properties added using `Object.defineProperty()` are not writable, not enumerable, and not configurable. In addition, `Object.defineProperty()` uses the [`[[DefineOwnProperty]]`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/Proxy/defineProperty) internal method, instead of [`[[Set]]`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/Proxy/set), so it does not invoke [setters](/en-US/docs/Web/JavaScript/Reference/Functions/set), even when the property is already present.
```

estelle committed:
commas

Update files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md

estelle commented on files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:
```diff
@@ -30,33 +28,21 @@ Object.defineProperty(obj, prop, descriptor)
-value, which may or may not be writable. An **accessor descriptor** is a
-property described by a getter-setter pair of functions. A descriptor must be one of
-these two flavors; it cannot be both.
+Property descriptors present in objects come in two main flavors: data descriptors and accessor descriptors. A **data descriptor** is a property that has a value, which may or may not be writable. An **accessor descriptor** is a property described by a getter-setter pair of functions. A descriptor must be one of these two flavors; it cannot be both.
```
```suggestion
Property descriptors present in objects come in two main flavors: data descriptors and accessor descriptors. A **data descriptor** is a property with a value that may or may not be writable. An **accessor descriptor** is a property described by a getter-setter pair of functions. A descriptor must be one of these two flavors; it cannot be both.
```

estelle commented on files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:
```diff
@@ -69,82 +55,61 @@ properties using `Object.defineProperty()`):
-  "Object.create(null)")}}.
+  - : A function which serves as a setter for the property, or {{jsxref("undefined")}} if there is no setter. When the property is assigned, this function is called with one argument (the value being assigned to the property) and with `this` set to the object through which the property is assigned. **Defaults to {{jsxref("undefined")}}.**
+
+If a descriptor has none of `value`, `writable`, `get` and `set` keys, it is treated as a data descriptor. If a descriptor has both \[`value` or `writable`] and \[`get` or `set`] keys, an exception is thrown.
```
```suggestion
If a descriptor has no `value`, `writable`, `get` or `set` key, it is treated as a data descriptor. If a descriptor has both \[`value` or `writable`] and \[`get` or `set`] keys, an exception is thrown.
```

Josh-Cena replied:
I find the suggestion slightly less precise since it sounds like "the descriptor is missing _any of_ those keys" instead of "_all of_ those keys"...

Elchi3 replied:
Would this capture it more clearly? I think the point is that "none of" negations don't read so easily. 

```suggestion
If a descriptor doesn't use a `value`, `writable`, `get` nor a `set` key, it is treated as a data descriptor. If a descriptor has both \[`value` or `writable`] and \[`get` or `set`] keys, an exception is thrown.
```

Josh-Cena replied:
```suggestion
If a descriptor doesn't have any of the `value`, `writable`, `get`, and `set` keys, it is treated as a data descriptor. If a descriptor has both \[`value` or `writable`] and \[`get` or `set`] keys, an exception is thrown.
```

Elchi3 replied:
Yes, I think that's great! 👍

estelle commented on files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:
```diff
@@ -69,82 +55,61 @@ properties using `Object.defineProperty()`):
+
+If a descriptor has none of `value`, `writable`, `get` and `set` keys, it is treated as a data descriptor. If a descriptor has both \[`value` or `writable`] and \[`get` or `set`] keys, an exception is thrown.
+
+Bear in mind that these attributes are not necessarily the descriptor's own properties. Inherited properties will be considered as well. In order to ensure these defaults are preserved, you might freeze existing objects in the descriptor object's prototype chain upfront, specify all options explicitly, or point to [`null`](/en-US/docs/Web/JavaScript/Reference/Operators/null) with {{jsxref("Object ...
```
```suggestion
Remember, these attributes are not necessarily the descriptor's own properties. Inherited properties will be considered as well. In order to ensure these defaults are preserved, you might freeze existing objects in the descriptor object's prototype chain upfront, specify all options explicitly, or point to [`null`](/en-US/docs/Web/JavaScript/Reference/Operators/null) with {{jsxref("Object.create", "Object.create(null)")}}.
```

Josh-Cena replied:
```suggestion
These attributes are not necessarily the descriptor's own properties. Inherited properties will be considered as well. In order to ensure these defaults are preserved, you might freeze existing objects in the descriptor object's prototype chain upfront, specify all options explicitly, or point to [`null`](/en-US/docs/Web/JavaScript/Reference/Operators/null) with {{jsxref("Object.create", "Object.create(null)")}}.
```

estelle commented on files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:
```diff
@@ -162,41 +127,45 @@ When the current property is configurable, defining an attribute to `undefined`
-When the property specified doesn't exist in the object,
-`Object.defineProperty()` creates a new property as described. Fields may be
-omitted from the descriptor, and default values for those fields are inputted.
+When the property specified doesn't exist in the object, `Object.defineProperty()` creates a new property as described. Fields may be omitted from the descriptor, and default values for those fields are inputted.
```
```suggestion
When the property specified doesn't exist in the object, `Object.defineProperty()` creates a new property as described. Fields may be omitted from the descriptor and default values for those fields are inputted.
```

estelle commented on files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:
```diff
@@ -209,15 +178,14 @@ When modifying an existing property, the current property configuration determin
 
-When the `writable` property attribute is set to `false`, the
-property is said to be "non-writable". It cannot be reassigned.
+When the `writable` property attribute is set to `false`, the property is said to be "non-writable". It cannot be reassigned. Trying to write into the non-writable property doesn't change it, and results in an error in [strict mode](/en-US/docs/Web/JavaScript/Reference/Strict_mode).
```
```suggestion
When the `writable` property attribute is `false`, the property is said to be "non-writable". It cannot be reassigned. Trying to write to a non-writable property doesn't change it and results in an error in [strict mode](/en-US/docs/Web/JavaScript/Reference/Strict_mode).
```

estelle commented on files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:
```diff
@@ -239,39 +207,31 @@ console.log(o.a); // 37; the assignment didn't work
-For non-{{jsxref("Global_Objects/Symbol", "Symbol")}} properties it also defines whether it shows
-up in a {{jsxref("Statements/for...in", "for...in")}} loop and
-{{jsxref("Object.keys()")}} or not.
+The `enumerable` property attribute defines whether the property is picked by {{jsxref("Object.assign()")}} or [spread](/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) operator. For non-{{jsxref("Global_Objects/Symbol", "Symbol")}} properties, it also defines whether it shows up in a {{jsxref("Statements/for...in", "for...in")}} loop and {{jsxref("Object.keys()")}} or not. For more  ...
```
I don't know what the term "picked" means here.  Maybe "selected" or "updated". I realize this was in the original. Maybe it's intuitive and this comment can be ignored, but it isn't clear to me.

Elchi3 replied:
"considered"?

```suggestion
The `enumerable` property attribute defines whether the property is considered by {{jsxref("Object.assign()")}} or the [spread](/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) operator. For non-{{jsxref("Global_Objects/Symbol", "Symbol")}} properties, it also defines whether it shows up in a {{jsxref("Statements/for...in", "for...in")}} loop and {{jsxref("Object.keys()")}} or not. For more information, see [Enumerability and ownership of properties](/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties).
```

estelle commented on files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:
```diff
@@ -281,128 +241,163 @@ for (const i in o) {
-The `configurable` attribute controls at the same time whether the property
-can be deleted from the object and whether its attributes (other than `value`
-and `writable`) can be changed.
+The `configurable` attribute controls at the same time whether the property can be deleted from the object and whether its attributes (other than `value` and `writable`) can be changed.
```
```suggestion
The `configurable` attribute controls whether the property can be deleted from the object and whether its attributes (other than `value` and `writable`) can be changed.
```

Not sure?

estelle commented on files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:
```diff
@@ -281,128 +241,163 @@ for (const i in o) {
 
 console.log(o.a); // 1
-delete o.a; // Nothing happens
+delete o.a; // Nothing happens; throws in strict mode
```
```suggestion
delete o.a; // Nothing happens; throws an error in strict mode
```

estelle commented on files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:
```diff
@@ -281,128 +241,163 @@ for (const i in o) {
+});
+Object.defineProperty(o, "b", {
+  value: 1,
+}); // TypeError: because the property is neither writable nor configurable, it cannot be modified
```
```suggestion
}); // TypeError: because the property is neither writable nor configurable; it cannot be modified
```

Josh-Cena replied:
```suggestion
}); // TypeError: the property cannot be modified because it is neither writable nor configurable, 
```

estelle commented on files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:
```diff
@@ -489,18 +477,15 @@ a.x = 1;
-Unlike accessor properties, value properties are always set on the object itself, not
-on a prototype. However, if a non-writable value property is inherited, it still
-prevents from modifying the property on the object.
+Unlike accessor properties, value properties are always set on the object itself, not on a prototype. However, if a non-writable value property is inherited, it still prevents from modifying the property on the object.
```
```suggestion
Unlike accessor properties, value properties are always set on the object itself, not on a prototype. However, if a non-writable value property is inherited, it is still prevented from modifying the property on the object.
```

Josh-Cena replied:
```suggestion
Unlike accessor properties, value properties are always set on the object itself, not on a prototype. However, if a non-writable value property is inherited, it is still prevented from being modified on the object.
```

estelle reviewed (commented):
A few minor grammar suggestions and one word clarification.

Josh-Cena committed:
Apply suggestions from code review

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

Josh-Cena committed:
Apply suggestions from code review

Elchi3 reviewed (approved):
Looks good to me modulo the three outstanding suggestions Estelle has made.

Josh-Cena committed:
Apply suggestions from code review

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>
Co-authored-by: Florian Scholz <fs@florianscholz.com>

Josh-Cena commented:
I've addressed all comments, sans one which I think the original is better

Elchi3 committed:
Update files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>
